### ISSUE TITLE
dhcp onboot causes startup to hang

### ISSUE THREAD

**Issue by deitch** [linuxkit/linuxkit#2226]
**Description**
If there is no NIC, or the network is disconnected, dhcpd hangs. If dhcpd is in `onboot` (as it is in the default `linuxkit.yml` and others), it blocks everything else afterwards, including `containerd`.

**Steps to reproduce the issue:**

1. Build the default `linuxkit.yml`
2. Run it on a host (virtual or physical) with a disconnected NIC
3. See that everything hangs

**Describe the results you received:**

Hang

**Describe the results you expected:**

No hang

**Additional information you deem important (e.g. issue happens only occasionally):**

No cute animals for issues? Maybe ones in the vet hospital, to indicate man issue? :-)

**Comment by justincormack** [comment]
Yes, that is expected. Don't run it as `onboot` if you might not have connectivity - this setup is really only meant for cloud or desktop setups. There is no option to `dhcpcd` to eventually time out. Obviously if you move to a service anything that needs networking will have to have its own "wait for networking" script if necessary, although presumably you are using code that is tolerant of no networking in this situation.

**Comment by deitch** [comment]
> although presumably you are using code that is tolerant of no networking in this situation.

Yeah, in this case just testing kernel configs for local hardware, but in general, I obviously want dhcp to work.

> Don't run it as onboot if you might not have connectivity

I can see that, but blocking `containerd` entirely?

In this case, it found my physical `eth0` but not the `wlan0`. I assume some drivers missing.

**Comment by justincormack** [comment]
We have no wireless support. It needs firmware etc.

We used to only support `dhcp` as a service, and only switched it as it is simpler for tests (which are largely onboot only) and some environments, but a large number of use cases should use it as a service (or both). In particular the one shot case is not suitable if your IP address might change either.

**Comment by deitch** [comment]
Yeah, I recall an issue: #1702 

So...

1. everything in `onboot` blocks, and all of it before `containerd`
2. `dhcpd` is in `onboot` if you *must* have it before other services, but then expect that if it fails, everything fails
3. Eventually `dhcpd` will be in both, or a "run-me-first" service.

That about it?

**Comment by justincormack** [comment]
No, 3 is not the case. You will still not put it on `onboot` if you have unreliable networking and you want services to run before networking is up. Most applications should work fine if it is a service, but it is terrible for running tests for example.
